service/history/replication/cache.go

```diff
@@ -0,0 +1,150 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2022 Uber Technologies Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package replication
+
+import (
+	"container/heap"
+	"errors"
+	"sync"
+
+	"github.com/uber/cadence/common/dynamicconfig"
+	"github.com/uber/cadence/common/types"
+)
+
+var (
+	errCacheFull    = errors.New("cache is full")
+	errAlreadyAcked = errors.New("already acked")
+)
+
+// Cache is an in-memory implementation of a cache for storing hydrated replication messages.
+// Messages can come out of order as long as their task ID is higher than last acknowledged message.
+// Out of order is expected as different source clusters will share hydrated replication messages.
+//
+// Cache utilizes heap to keep replication messages in order. This is needed for efficient acknowledgements in O(log N).
+//
+// Cache capacity can be increased dynamically. Decrease will require a restart, as new tasks will not be accepted, but memory will not be reclaimed either.
+//
+// Cache methods are thread safe. It is expected to have writers and readers from different go routines.
+type Cache struct {
```

Comment:
Looks like it's worth making it an interface, considering we probably will introduce disk cache later.
But it can be done later, not necessarily in this PR

Reply:
Generally in GoLang (as opposed to some other languages) we should avoid introducing interfaces until we actually need them. It is also a good practice to define them in consumer package instead (unless they are very common interfaces, used in multiple places).

https://github.com/golang/go/wiki/CodeReviewComments#interfaces
https://bryanftan.medium.com/accept-interfaces-return-structs-in-go-d4cab29a301b